### PR TITLE
Merge Epsilon constants into 1

### LINKED ISSUE #1023: Geom: Only have one use of EPSILON
### Description of the task

We have a couple of different definitions of ```EPSILON``` such as ```EPS``` and even ```EPS2```. We should change any usages of such definitions and only have one ```EPSILON``` (the one that lives in ```GeomConstants.h```. 

### Acceptance criteria

- [ ] Only have one ```EPSILON``` being used throughout our codebase
- [ ] All tests that depended on such usages still pass

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description

Removed instances of EPS and EPS2 found and replaced with `FIXED_EPSILON` found in `geom_constants.h`.

### Testing Done

Tests are still passing.

### Resolved Issues

Resolves #1023

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by kelvinkoon** [comment]
Was wondering if ALL instances of EPSILON (i.e 1e-15) should be replaced to use `FIXED_EPSILON` found in `geom_constants.h`. For instance, `time.h` has its own public member of EPSILON. Would we want that to be `FIXED_EPSILON` as well?

**Comment by MathewMacDougall** [comment]
> Was wondering if ALL instances of EPSILON (i.e 1e-15) should be replaced to use FIXED_EPSILON found in geom_constants.h. For instance, time.h has its own public member of EPSILON. Would we want that to be FIXED_EPSILON as well?

IIRC the reason we originally had separate `EPSILON` values per-class was in-case they needed to be different. For things like `Time` I think it's fine to leave as-is, otherwise `geom_constants.h` should be moved/named something else (since Time isn't geom).

@garethellis0 @jonathanlew @akhilveeraghanta thoughts?

**Comment by kelvinkoon** [comment]
> > Was wondering if ALL instances of EPSILON (i.e 1e-15) should be replaced to use FIXED_EPSILON found in geom_constants.h. For instance, time.h has its own public member of EPSILON. Would we want that to be FIXED_EPSILON as well?
> 
> IIRC the reason we originally had separate `EPSILON` values per-class was in-case they needed to be different. For things like `Time` I think it's fine to leave as-is, otherwise `geom_constants.h` should be moved/named something else (since Time isn't geom).
> 
> @garethellis0 @jonathanlew @akhilveeraghanta thoughts?

Alright, I'll check with grep again later. CI is failing due to a test in `calc_best_shot.cpp` so I'll have to look at that anyways.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

It's worth noting that all these epsilon values will be going away with `new_geom` as we've moved to a more accurate / nuanced view of what "close" values are.

**Comment by jonathanlew** [comment]
> LGTM +1 tada
> 
> It's worth noting that all these epsilon values will be going away with `new_geom` as we've moved to a more accurate / nuanced view of what "close" values are.

Is there a ticket for that?

**Comment by jonathanlew** [comment]
> > Was wondering if ALL instances of EPSILON (i.e 1e-15) should be replaced to use FIXED_EPSILON found in geom_constants.h. For instance, time.h has its own public member of EPSILON. Would we want that to be FIXED_EPSILON as well?
> 
> IIRC the reason we originally had separate `EPSILON` values per-class was in-case they needed to be different. For things like `Time` I think it's fine to leave as-is, otherwise `geom_constants.h` should be moved/named something else (since Time isn't geom).
> 
> @garethellis0 @jonathanlew @akhilveeraghanta thoughts?

Leave Time as is; Time is totally unrelated to geom

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI

**Review by MathewMacDougall — DISMISSED** [review]
LGTM pending CI

**Comment by kelvinkoon** [comment]
Did a bit of poking and found the culprit to be in `util.cpp`'s `getEmptySpaceWithinParentSegment` function. Specifically, when the function tries to remove all zero length open segments, there is a check with the `length()` method that if it is less than EPSILON, it'll be erased.
Before this ticket, it was using a value of `1e-9`, whereas replacing it with the `geom_constants.h`'s `1e-15` will cause the `test_calc_open_shot_with_a_dense_wall_of_obstacles_2` test in `calc_best_shot_test.cpp` to fail.
Any preferred ways on how we want to deal with this?

**Comment by jonathanlew** [comment]
> 
> 
> Did a bit of poking and found the culprit to be in `util.cpp`'s `getEmptySpaceWithinParentSegment` function. Specifically, when the function tries to remove all zero length open segments, there is a check with the `length()` method that if it is less than EPSILON, it'll be erased.
> Before this ticket, it was using a value of `1e-9`, whereas replacing it with the `geom_constants.h`'s `1e-15` will cause the `test_calc_open_shot_with_a_dense_wall_of_obstacles_2` test in `calc_best_shot_test.cpp` to fail.
> Any preferred ways on how we want to deal with this?

Oh, why is it failing? Can you dig in to what the test is doing and whether something about the function or the test should be changed?

**Comment by kelvinkoon** [comment]
> > 
> > 
> > Did a bit of poking and found the culprit to be in `util.cpp`'s `getEmptySpaceWithinParentSegment` function. Specifically, when the function tries to remove all zero length open segments, there is a check with the `length()` method that if it is less than EPSILON, it'll be erased.
> > Before this ticket, it was using a value of `1e-9`, whereas replacing it with the `geom_constants.h`'s `1e-15` will cause the `test_calc_open_shot_with_a_dense_wall_of_obstacles_2` test in `calc_best_shot_test.cpp` to fail.
> > Any preferred ways on how we want to deal with this?
> 
> Oh, why is it failing? Can you dig in to what the test is doing and whether something about the function or the test should be changed? 

The test is presumably failing because the tolerance changed and it's not erasing a segment due to it falling between 1e-9 and 1e-15.

I would say it's probably better to change the test since the function looks like it's doing what it should be.

**Comment by MathewMacDougall** [comment]
@EvanMorcom perhaps we change calcOpenShot again? :thinking: :eyes: :fire_extinguisher: (jk-ish)

**Comment by garethellis0** [comment]
> > LGTM +1 tada
> > It's worth noting that all these epsilon values will be going away with `new_geom` as we've moved to a more accurate / nuanced view of what "close" values are.
> 
> Is there a ticket for that?

There's not currently. We should probably do a pass through all the geom tickets, remove old ones and add any relevant new ones. I think several are out of date.

**Comment by akhilveeraghanta** [comment]
@kelvinkoon please merge upstream/master to fix CI

**Comment by kelvinkoon** [comment]
Checked out `getEmptySpaceWithinParentSegment()` and the `open_segs` has 2 segments. The segments both start and end at the same place, such as (4.5, -0.15) to (4.5, -0.15). 
The open segment lengths are `1.77e-15` and `2.77e-15`. `1.77e-15` < `FIXED_EPSILON` (1e-15) evaluates to false. 
To attempt to resolve this, I changed FIXED_EPSILON to `2e-15`.

File: src/software/new_geom/geom_constants.h [hunk 1]
```diff
@@ -6,7 +6,7 @@
 // needs of operations near such scale. This is not meant to cover all use cases, and
 // other epsilon values may need to be chosen for specific use cases.
 // Further reading: https://bitbashing.io/comparing-floats.html
-static constexpr double FIXED_EPSILON = 1e-15;
+static constexpr double FIXED_EPSILON = 2e-15;
```
Comment (jonathanlew) [inline src/software/new_geom/geom_constants.h]:
```suggestion
static constexpr double FIXED_EPSILON = 1e-15;
```

**Review by jonathanlew — CHANGES_REQUESTED** [review]
We chose 1e-15 for a good reason. If we need 2e-15 for a geom application, then we can do 2\*FIXED_EPSILON and comment why we need that 2\* factor

File: src/software/new_geom/geom_constants.h (hunk 1 above)
Comment (kelvinkoon) [reply src/software/new_geom/geom_constants.h]: Reverted.

**Review by akhilveeraghanta — DISMISSED** [review]
Lgtm 💯

File: src/software/geom/util.cpp [hunk 2]
```diff
@@ -507,7 +508,10 @@ std::vector<Segment> getEmptySpaceWithinParentSegment(std::vector<Segment> segme
     for (std::vector<Segment>::const_iterator it = open_segs.begin();
          it != open_segs.end();)
     {
-        if (it->length() < EPS)
+        // If segment length is less than 2 * FIXED_EPSILON, remove the segment
+        // 2 * FIXED_EPSILON chosen as FIXED_EPSILON by itself is not a sufficient for all
+        // cases
```
Comment (jonathanlew) [inline src/software/geom/util.cpp]: Sounds like a tautological justification. Can we investigate this issue to see if there are reasons why we're accumulating so much error? For example, two floating point operations would mean we could expect up to 2 \* FIXED_EPSILON error. See test_util.cpp:equalWithinTolerance for some examples.

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/geom/util.cpp (hunk 2 above)
Comment (kelvinkoon) [reply src/software/geom/util.cpp]: Improved comment, let me know if there's any issues in the reasoning.

**Review by jonathanlew — APPROVED** [review]
:+1:

**Review by LiCody — APPROVED** [review]
LGTM 🎉

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Pull request merged**